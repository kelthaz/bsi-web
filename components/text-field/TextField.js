import styles from './text-field.module.scss';

export const TextField = (props) => {
  return <div className={styles.group}>
    <input className={styles.input} type="text" required/>
    <span className={styles.highlight}></span>
    <span className={styles.bar}></span>
    <label className={`body2 ${styles.label}`}>{props.label}</label>
  </div>;
};

export default TextField;
