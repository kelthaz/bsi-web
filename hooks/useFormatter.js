import phoneFormatter from '../helpers/phoneFormatter';
import uppercaseFormatter from '../helpers/uppercaseFormatter';

const useFormatter = (format) => {
  switch (format) {
    case 'phone':
      return { formatter: phoneFormatter, changeSelection: true };

    case 'uppercase':
      return { formatter: uppercaseFormatter, changeSelection: true };

    default:
      return (value) => value;
  }
};

export default useFormatter;
