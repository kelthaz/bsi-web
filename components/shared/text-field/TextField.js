/* eslint-disable complexity */
import PropTypes from 'prop-types';
import { useState } from 'react';
import SvgCheckOk from '../../svgs/SvgCheckOk';
import SvgCross from '../../svgs/SvgCross';
import SvgHidenPassword from '../../svgs/SvgHidenPassword';
import SvgShowPassword from '../../svgs/SvgShowPassword';
import styles from './text-field.module.scss';
import useFormatter from '../../../hooks/useFormatter';

const seleccionaEstilo = (size, inverted) => {
  const finalStyles = [];
  if (inverted) {
    finalStyles.push(size === 'big' ? styles['input-big-inverted'] : styles['input-small-inverted']);
    finalStyles.push(styles['icon-check-inverted']);
    finalStyles.push(styles['label-small-inverted']);
    finalStyles.push(styles['indicador-activo-inverted']);
    finalStyles.push(styles['help-text-inverted']);
  } else {
    finalStyles.push(size === 'big' ? styles['input-big'] : styles['input-small']);
    finalStyles.push(styles['icon-check']);
    finalStyles.push(styles['label-small']);
    finalStyles.push(styles['indicador-activo']);
    finalStyles.push(styles['help-text']);
  }
  return finalStyles;
};

const TextField = (props) => {
  const {
    name,
    formulario,
    maxlength,
    capitalize,
    label,
    type: typeInput,
    size,
    inverted,
    optional,
    validation,
    format,
    paste,
    readonly,
    disabled,
  } = props;
  const [inputStyle, iconCheckStyle, labelStyle, indicadorStyle, helpTextStyle] = seleccionaEstilo(size, inverted);
  const { handleChange, values, handleBlur, errors, touched, setFieldTouched, setFieldValue } = formulario;

  const [type, setType] = useState(typeInput);

  const error = <SvgCross className={styles['icon-error']} />;
  const status = <SvgCheckOk className={iconCheckStyle} />;

  const { formatter, changeSelection, changeSelectionFunc } = useFormatter(format);
  const [active, setActive] = useState(false);

  let keyPress = '';
  let pasteCurrent = false;

  const beforeInput = (event) => {
    if (!pasteCurrent && event.data.length === 2) {
      event.preventDefault();
    }
    // setFieldValue(name, event.data);
  };

  const onHandleChange = (event) => {
    if (!touched[name]) {
      setFieldTouched(name, true);
    }
    const { selectionStart, selectionEnd } = changeSelectionFunc(event, keyPress);

    event.target.value = formatter(event.target.value.trimStart());
    handleChange(event);

    if (changeSelection) {
      event.target.setSelectionRange(selectionStart, selectionEnd);
    }

    keyPress = '';
    pasteCurrent = false;
  };

  const onHandleBlur = (event) => {
    handleBlur(event);
    setActive(false);
  };

  const onPaste = (event) => {
    if (!paste) {
      event.preventDefault();
    } else {
      pasteCurrent = true;
    }
  };

  const indicadorError = type === 'password' ? styles['indicador-error-password'] : styles['indicador-error'];
  const hasError = () => touched[name] && errors[name];

  return (
    <div className={`${styles.group}`}>
      <input
        id={name}
        name={name}
        className={`${type === 'password' ? styles['input-big-password'] : inputStyle} ${
          capitalize ? styles.capitalize : ''
        } ${hasError() ? indicadorError : active && indicadorStyle} `}
        type={type}
        onChange={onHandleChange}
        onBlur={(event) => {
          onHandleBlur(event);
          setFieldValue(name, values[name].trim());
        }}
        value={values[name]}
        maxLength={maxlength}
        autoComplete="off"
        placeholder={size === 'big' ? label : ''}
        onFocus={() => {
          setActive(true);
        }}
        onPaste={onPaste}
        onKeyDown={(event) => {
          keyPress = event.key;
        }}
        readOnly={readonly}
        tabIndex="0"
        disabled={disabled}
        onBeforeInput={beforeInput}
      />

      {size === 'small' && (
        <label
          htmlFor={name}
          className={`${disabled ? styles['label-disabled'] : labelStyle} ${hasError() ? styles['label-error'] : ''} ${
            values[name] !== '' || active ? styles['label-active'] : ''
          }`}
        >
          {label}
        </label>
      )}

      <span className={hasError() ? styles['help-text-error'] : helpTextStyle}>
        {hasError() ? errors[name] : active && optional && 'Opcional'}&nbsp;
      </span>
      {typeInput === 'password' && (
        <button
          className={styles['button-password-inverted']}
          type="button"
          onClick={() => setType(type === 'password' ? 'text' : 'password')}
        >
          {type === 'text' ? <SvgShowPassword /> : <SvgHidenPassword />}
        </button>
      )}
      {hasError() ? (
        <div className={typeInput === 'password' ? styles['status-icon-password'] : styles['status-icon']}>{error}</div>
      ) : (
        validation && <div className={styles['status-icon']}>{active && status}</div>
      )}
    </div>
  );
};

TextField.propTypes = {
  capitalize: PropTypes.bool,
  label: PropTypes.any.isRequired,
  name: PropTypes.string.isRequired,
  formulario: PropTypes.any.isRequired,
  type: PropTypes.string.isRequired,
  size: PropTypes.string.isRequired,
  inverted: PropTypes.bool,
  optional: PropTypes.bool,
  validation: PropTypes.bool,
  paste: PropTypes.bool,
  maxlength: PropTypes.number.isRequired,
  format: PropTypes.string,
  readonly: PropTypes.bool,
  disabled: PropTypes.bool,
};

TextField.defaultProps = {
  capitalize: false,
  inverted: false,
  optional: false,
  validation: false,
  paste: true,
  readonly: false,
  disabled: false,
  format: '',
};

export default TextField;
