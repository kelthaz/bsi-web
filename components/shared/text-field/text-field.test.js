import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import React from 'react';
import { act } from 'react-dom/test-utils';
import useFormatter from '../../../hooks/useFormatter';
import TextField from './TextField';

jest.mock('../../../hooks/useFormatter', () => jest.fn());

describe('Pruebas en el componente TextField', () => {
  const setValue = jest.fn();
  const setTouched = jest.fn();
  const changeSelection = jest.fn();

  beforeEach(() => {
    useFormatter.mockReturnValue([jest.fn((value) => value), changeSelection]);
    jest.clearAllMocks();
  });

  it('Debe de mostrarse correctamente el componente', () => {
    // arrange
    const props = {
      name: 'name',
      format: 'uppercase',
      touched: false,
      value: '',
      setValue,
      setTouched,
      size: 'small',
      capitalize: true,
      label: 'Nombre',
      type: 'text',
      maxlength: 60,
      inverted: true,
    };
    // act
    const wrapper = shallow(<TextField {...props} />);
    // assert
    expect(wrapper).toMatchSnapshot();
  });

  it('Debe de llamarse el setValue y el setTouched, con el valor nuevo', async () => {
    // arrange
    const valorEsperado = 'JESUS';
    const props = {
      name: 'name',
      format: 'uppercase',
      touched: false,
      value: '',
      setValue,
      setTouched,
      size: 'big',
      capitalize: false,
      label: 'Nombre',
      type: 'text',
      maxlength: 60,
      inverted: false,
    };
    const wrapper = shallow(<TextField {...props} />);
    // act
    await act(async () => {
      wrapper
        .find('input')
        .simulate('change', { target: { value: valorEsperado, selectionStart: 0, setSelectionRange: jest.fn() } });
    });

    // assert
    expect(setValue).toHaveBeenCalledTimes(1);
    expect(setValue).toHaveBeenCalledWith(valorEsperado);
    expect(setTouched).toHaveBeenCalledTimes(1);
    expect(setTouched).toHaveBeenCalledWith(true);
  });

  test('Debe de llamarse el setValue y no el setTouched, con el valor nuevo', async () => {
    // arrange
    const valorEsperado = 'soy una empresa que vende empanadas en cali';
    const props = {
      name: 'name',
      format: 'uppercase',
      touched: true,
      value: '',
      setValue,
      setTouched,
      size: 'big',
      capitalize: false,
      label: 'Nombre',
      type: 'password',
      maxlength: 60,
      inverted: false,
    };
    const wrapper = shallow(<TextField {...props} />);
    // act
    await act(async () => {
      wrapper
        .find('input')
        .simulate('change', { target: { value: valorEsperado, selectionStart: 0, setSelectionRange: jest.fn() } });
    });
    // assert
    expect(setValue).toHaveBeenCalledTimes(1);
    expect(setValue).toHaveBeenCalledWith(valorEsperado);
    expect(setTouched).toHaveBeenCalledTimes(0);
  });

  test('Debe de llamarse el setValue y no el setTouched, con el espacio final eliminado', async () => {
    // arrange
    const valorEsperado = 'JESUS ';
    const props = {
      name: 'name',
      format: 'uppercase',
      touched: true,
      value: valorEsperado,
      setValue,
      setTouched,
      size: 'big',
      capitalize: false,
      label: 'Nombre',
      type: 'password',
      maxlength: 60,
      inverted: false,
    };
    const wrapper = shallow(<TextField {...props} />);
    // act
    await act(async () => {
      wrapper.find('input').simulate('blur');
    });
    // assert
    expect(setValue).toHaveBeenCalledTimes(1);
    expect(setValue).toHaveBeenCalledWith(valorEsperado.trimEnd());
    expect(setTouched).toHaveBeenCalledTimes(0);
  });

  test('Debe de llamarse el preventDefault porque el valor ingresado no es valido', () => {
    // arrange
    useFormatter.mockReturnValue([jest.fn(() => ''), changeSelection]);
    const preventDefault = jest.fn();
    const props = {
      name: 'name',
      format: 'uppercase',
      touched: true,
      value: '',
      setValue,
      setTouched,
      size: 'big',
      capitalize: false,
      label: 'Nombre',
      type: 'password',
      maxlength: 60,
      inverted: false,
    };
    const wrapper = shallow(<TextField {...props} />);
    // act
    wrapper.find('input').simulate('beforeInput', { data: '$', preventDefault });
    // assert
    expect(preventDefault).toHaveBeenCalledTimes(1);
  });

  test('Debe de no llamarse el preventDefault porque el valor ingresado es valido', () => {
    // arrange
    const preventDefault = jest.fn();
    const props = {
      name: 'name',
      format: 'uppercase',
      touched: true,
      value: '',
      setValue,
      setTouched,
      size: 'big',
      capitalize: false,
      label: 'Nombre',
      type: 'text',
      maxlength: 60,
      inverted: false,
    };
    const wrapper = shallow(<TextField {...props} />);
    // act
    wrapper.find('input').simulate('beforeInput', { data: 'oli', preventDefault });
    // assert
    expect(preventDefault).toHaveBeenCalledTimes(0);
  });

  test('Debe de mostrarse en rojo debido al error', () => {
    // arrange
    const error = 'Campo obligatorio';
    const props = {
      name: 'name',
      format: 'uppercase',
      touched: true,
      value: '',
      error,
      setValue,
      setTouched,
      size: 'big',
      capitalize: false,
      label: 'Nombre',
      type: 'text',
      maxlength: 60,
      inverted: false,
    };
    const wrapper = shallow(<TextField {...props} />);
    // act
    const indicadorError = wrapper.find('.opacity-error').exists();
    const errorMostrado = wrapper.find('span').text().trim();
    // assert
    expect(indicadorError).toBe(true);
    expect(errorMostrado).toBe(error);
  });

  test('Debe de mostrarse como opcional', () => {
    // arrange
    const props = {
      name: 'name',
      format: 'uppercase',
      touched: true,
      value: '',
      setValue,
      setTouched,
      size: 'big',
      capitalize: false,
      label: 'Nombre',
      type: 'password',
      maxlength: 60,
      inverted: false,
      optional: true,
    };
    const wrapper = shallow(<TextField {...props} />);
    // act
    const opcional = wrapper.find('span').text().trim();
    // assert
    expect(opcional).toBe('Opcional');
  });

  test('Debe de presionarse la tecla Delete', async () => {
    // arrange
    const key = 'Delete';
    const target = { value: key, selectionStart: 0, setSelectionRange: jest.fn() };
    const props = {
      name: 'name',
      format: 'uppercase',
      touched: false,
      value: '',
      setValue,
      setTouched,
      size: 'big',
      capitalize: false,
      label: 'Nombre',
      type: 'password',
      maxlength: 60,
      inverted: false,
      optional: true,
    };
    const wrapper = shallow(<TextField {...props} />);
    // act
    wrapper.find('input').simulate('keyDown', { key });
    await act(async () => {
      wrapper.find('input').simulate('change', { target });
    });
    // assert
    expect(setValue).toHaveBeenCalledTimes(1);
    expect(setValue).toHaveBeenCalledWith(key);
    expect(setTouched).toHaveBeenCalledTimes(1);
    expect(setTouched).toHaveBeenCalledWith(true);
    expect(changeSelection).toHaveBeenCalledTimes(1);
    expect(changeSelection).toHaveBeenCalledWith(key, props.value, key, key, 0, target);
  });

  test('Debe de mostrar el texto en el campo tipo password', () => {
    // arrange
    const props = {
      name: 'name',
      format: 'uppercase',
      touched: true,
      value: '',
      setValue,
      setTouched,
      size: 'big',
      capitalize: false,
      label: 'Nombre',
      type: 'password',
      maxlength: 60,
      inverted: false,
    };
    const wrapper = shallow(<TextField {...props} />);
    // act
    wrapper.find('button').simulate('click');
    const showPassword = wrapper.find('SvgShowPassword').exists();
    // assert
    expect(showPassword).toBe(true);
  });

  test('Debe de mostrar el texto y luego ocultarlo de nuevo en el campo tipo password', () => {
    // arrange
    const props = {
      name: 'name',
      format: 'uppercase',
      touched: true,
      value: '',
      setValue,
      setTouched,
      size: 'big',
      capitalize: false,
      label: 'Nombre',
      type: 'password',
      maxlength: 60,
      inverted: false,
    };
    const wrapper = shallow(<TextField {...props} />);
    // act
    wrapper.find('button').simulate('click');
    const showPassword = wrapper.find('SvgShowPassword').exists();
    wrapper.find('button').simulate('click');
    const hidenPassword = wrapper.find('SvgHidenPassword').exists();
    // assert
    expect(showPassword).toBe(true);
    expect(hidenPassword).toBe(true);
  });

  test('Debe de no llamarse el preventDefault porque permite pegar', () => {
    // arrange
    const preventDefault = jest.fn();
    const props = {
      name: 'name',
      format: 'uppercase',
      touched: true,
      value: '',
      setValue,
      setTouched,
      size: 'big',
      capitalize: false,
      label: 'Nombre',
      type: 'password',
      maxlength: 60,
      inverted: false,
      paste: true,
    };
    const wrapper = shallow(<TextField {...props} />);
    // act
    wrapper.find('input').simulate('paste');
    // assert
    expect(preventDefault).toHaveBeenCalledTimes(0);
  });

  test('Debe de llamarse el preventDefault porque permite no pegar', () => {
    // arrange
    const preventDefault = jest.fn();
    const props = {
      name: 'name',
      format: 'uppercase',
      touched: true,
      value: '',
      error: 'No se permite pegar',
      setValue,
      setTouched,
      size: 'big',
      capitalize: false,
      label: 'Nombre',
      type: 'password',
      maxlength: 60,
      inverted: false,
      paste: false,
    };
    const wrapper = shallow(<TextField {...props} />);
    // act
    wrapper.find('input').simulate('paste', { data: 'oli', preventDefault });
    // assert
    expect(preventDefault).toHaveBeenCalledTimes(1);
  });
});
