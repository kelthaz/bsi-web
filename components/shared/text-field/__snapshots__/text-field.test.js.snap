// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente TextField Debe de mostrarse correctamente el componente 1`] = `
<div
  className="group"
>
  <input
    autoComplete="off"
    className="input-small-inverted capitalize "
    disabled={false}
    id="name"
    maxLength={60}
    name="name"
    onBeforeInput={[Function]}
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    onPaste={[Function]}
    placeholder=" "
    readOnly={false}
    tabIndex="0"
    type="text"
    value=""
  />
  <label
    className="text-overflow"
    htmlFor="name"
  >
    Nombre
  </label>
  <span>
    Â 
  </span>
</div>
`;
